File: Algorithms/0073_Set_Matrix_Zeroes/Java/SetMatrixZeroesSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public void setZeroes(int[][] matrix) {
        if (matrix.length == 0) return;
        Set<Integer> row = new HashSet<>();
        Set<Integer> column = new HashSet<>();

        for (int y = 0; y < matrix.length; y++) {
            for (int x = 0; x < matrix[0].length; x++) {
                if (matrix[y][x] == 0) {
                    row.add(y);
                    column.add(x);
                }
            }
        }
        for (Integer i : row) {
            Arrays.fill(matrix[i], 0);
        }
        for (Integer i : column) {
            int temp = 0;
            while (temp < matrix.length) {
                matrix[temp][i] = 0;
                temp++;
            }
        }
    }
}

File: Algorithms/0073_Set_Matrix_Zeroes/Python/Set_Matrix_Zeroes_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def setZeroes(self, matrix):
        """
        Do not return anything, modify matrix in-place instead.
        """
        m = len(matrix)
        if m == 0: return
        n = len(matrix[0])
        cache = []
        for y in range(m):
            for x in range(n):
                if matrix[y][x] == 0:
                    cache.append([x, y])

        for x, y in cache:
            self.make_zero(x, y, m, n, matrix)

    def make_zero(self, x, y, m, n, matrix):
        index_x, index_y = 0, 0

        while index_x < n:
            matrix[y][index_x] = 0
            index_x += 1

        while index_y < m:
            matrix[index_y][x] = 0
            index_y += 1




